### ISSUE TITLE
device manager can incorrectly recover devices on node restart

### ISSUE THREAD

**Issue by ffromani** [kubernetes/kubernetes#109595]
### What happened?

The following behaviour was reproduced on openshift 4.10 ("Kubelet version" kubeletVersion="v1.23.3+54654d2") but I believe the root cause is on kubernetes.

Let's consider a pod requiring resources, among them a device (point in case: dpdk network device)
```
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: dpdk
  name: testpmd
  namespace: default
spec:
  containers:
    - command:
        - /bin/bash
        - -c
        - sleep INF
      image: $DPKDK_IMAGE
      imagePullPolicy: Always
      name: dpdk
      resources:
        limits:
          memory: 1000Mi
          hugepages-1Gi: 1Gi
          cpu: '2'
          openshift.io/dpdk_nic_1: 1
        requests:
          memory: 1000Mi
          hugepages-1Gi: 1Gi
          cpu: '2'
          openshift.io/dpdk_nic_1: 1
      securityContext:
        capabilities:
          add:
            - IPC_LOCK
            - SYS_RESOURCE
            - NET_RAW
            - NET_ADMIN
        runAsUser: 0
```
the pods starts and run just fine up until the node is restarted. On restart, the device manager correctly recover its state from the checkpoint data:
```
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.461850    2011 manager.go:247] "Starting Device Plugin manager"
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.463809    2011 container_manager_linux.go:930] "Attempting to apply oom_score_adj to process" oomScoreAdj=-999 pid=2011
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.464997    2011 oom_linux.go:66] attempting to set "/proc/2011/oom_score_adj" to "-999"
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.464269    2011 iptables.go:462] running iptables: ip6tables [-w 5 -W 100000 -N KUBE-POSTROUTING -t nat]
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.465254    2011 event.go:294] "Event occurred" object="TESTNODE" kind="Node" apiVersion="" type="Normal" reason="NodeAllocatableEnforced" message="Updated Node Allocatable limit across pods"
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.466379    2011 pod_devices.go:215] "Get checkpoint entry" podUID="c7dcb25a-6b71-4f80-9482-053ea597230c" containerName="dpdk" resourceName="openshift.io/dpdk_nic_1" deviceIDs=map[-1:[0000:00:05.0]] allocated="\n1\n!PCIDEVICE_OPENSHIFT_IO_DPDK_NIC_1\x12\f0000:00:05.0\x1a%\n\x0e/dev/vfio/vfio\x12\x0e/dev/vfio/vfio\x1a\x03mrw\x1a'\n\v/dev/vfio/0\x12\x13/dev/vfio/noiommu-0\x1a\x03mrw"
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.466823    2011 manager.go:289] "Serving device plugin registration server on socket" path="/var/lib/kubelet/device-plugins/kubelet.sock"
```
As expected, the device plugin capacity is set to zero until the device plugin correctly re-register itself:
```
Apr 13 09:53:14 TESTNODE hyperkube[2011]: I0413 09:53:14.607362    2011 setters.go:344] "Updated capacity for device plugin" plugin="openshift.io/dpdk_nic_1" capacity=0`
```
A bit later, the recovery process from the apiserver begins:
```
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193478    2011 config.go:278] "Setting pods for source" source="api"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193526    2011 config.go:383] "Receiving a new pod" pod="openshift-cluster-node-tuning-operator/tuned-tql4x"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193541    2011 config.go:383] "Receiving a new pod" pod="openshift-sdn/sdn-fnls4"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193549    2011 config.go:383] "Receiving a new pod" pod="openshift-monitoring/node-exporter-jzprg"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193557    2011 config.go:383] "Receiving a new pod" pod="openshift-ingress-canary/ingress-canary-62kq2"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193565    2011 config.go:383] "Receiving a new pod" pod="openshift-multus/multus-87sbt"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193573    2011 config.go:383] "Receiving a new pod" pod="openshift-multus/network-metrics-daemon-5qb5q"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193580    2011 config.go:383] "Receiving a new pod" pod="openshift-image-registry/node-ca-frkhr"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193597    2011 config.go:383] "Receiving a new pod" pod="openshift-monitoring/prometheus-adapter-689c895f8f-zpsx2"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193605    2011 config.go:383] "Receiving a new pod" pod="openshift-operator-lifecycle-manager/collect-profiles-27497370-6vd5b"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193615    2011 config.go:383] "Receiving a new pod" pod="openshift-operator-lifecycle-manager/collect-profiles-27497385-mwtfp"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193624    2011 config.go:383] "Receiving a new pod" pod="openshift-machine-config-operator/machine-config-daemon-nkqsq"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193632    2011 config.go:383] "Receiving a new pod" pod="openshift-image-registry/image-pruner-27496800-n4qzq"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193640    2011 config.go:383] "Receiving a new pod" pod="openshift-cluster-csi-drivers/openstack-cinder-csi-driver-node-qklm6"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193649    2011 config.go:383] "Receiving a new pod" pod="openshift-marketplace/246a7cd23747989b8f475c6ffc04f7e523236656eaac2249a6d4ddf03bvpl6x"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193659    2011 config.go:383] "Receiving a new pod" pod="openshift-multus/multus-additional-cni-plugins-xwgkj"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193667    2011 config.go:383] "Receiving a new pod" pod="openshift-openstack-infra/coredns-TESTNODE"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193674    2011 config.go:383] "Receiving a new pod" pod="openshift-dns/node-resolver-6n2vp"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193682    2011 config.go:383] "Receiving a new pod" pod="openshift-openstack-infra/keepalived-TESTNODE"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193689    2011 config.go:383] "Receiving a new pod" pod="openshift-dns/dns-default-cnlrl"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193695    2011 config.go:383] "Receiving a new pod" pod="default/testpmd"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193703    2011 config.go:383] "Receiving a new pod" pod="openshift-operator-lifecycle-manager/collect-profiles-27497355-cdncm"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193711    2011 config.go:383] "Receiving a new pod" pod="openshift-sriov-network-operator/sriov-network-config-daemon-p94vm"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193721    2011 config.go:383] "Receiving a new pod" pod="openshift-sriov-network-operator/sriov-device-plugin-bw7rq"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.193728    2011 config.go:383] "Receiving a new pod" pod="openshift-network-diagnostics/network-check-target-2tbjs"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.195257    2011 kubelet.go:2106] "SyncLoop ADD" source="api" pods=[openshift-cluster-node-tuning-operator/tuned-tql4x openshift-sdn/sdn-fnls4 openshift-monitoring/node-exporter-jzprg openshift-ingress-canary/ingress-canary-62kq2 openshift-multus/multus-87sbt openshift-multus/network-metrics-daemon-5qb5q openshift-image-registry/node-ca-frkhr openshift-monitoring/prometheus-adapter-689c895f8f-zpsx2 openshift-operator-lifecycle-manager/collect-profiles-27497370-6vd5b openshift-operator-lifecycle-manager/collect-profiles-27497385-mwtfp openshift-machine-config-operator/machine-config-daemon-nkqsq openshift-image-registry/image-pruner-27496800-n4qzq openshift-cluster-csi-drivers/openstack-cinder-csi-driver-node-qklm6 openshift-marketplace/246a7cd23747989b8f475c6ffc04f7e523236656eaac2249a6d4ddf03bvpl6x openshift-multus/multus-additional-cni-plugins-xwgkj openshift-openstack-infra/coredns-TESTNODE openshift-dns/node-resolver-6n2vp openshift-openstack-infra/keepalived-TESTNODE openshift-dns/dns-default-cnlrl default/testpmd openshift-operator-lifecycle-manager/collect-profiles-27497355-cdncm openshift-sriov-network-operator/sriov-network-config-daemon-p94vm openshift-sriov-network-operator/sriov-device-plugin-bw7rq openshift-network-diagnostics/network-check-target-2tbjs]
```
we can see already that the workload pod (testpmd) is going to be recovered BEFORE the device plugin. When the workload pod 
enters the admission flow (as per recovered pods), because the devicemanager restored its state and has data about this pod, we see:
```
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204837    2011 topology_manager.go:200] "Topology Admit Handler"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204848    2011 config.go:97] "Looking for sources, have seen" sources=[api file] seenSources=map[file:{}]
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204859    2011 topology_hints.go:49] "Resource does not have a topology preference" resource="openshift.io/dpdk_nic_1"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204873    2011 scope_container.go:77] "TopologyHints" hints=map[openshift.io/dpdk_nic_1:[]] pod="default/testpmd" containerName="dpdk"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204880    2011 config.go:97] "Looking for sources, have seen" sources=[api file] seenSources=map[file:{}]
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204928    2011 policy_static.go:437] "TopologyHints generated" pod="default/testpmd" containerName="dpdk" cpuHints=[{NUMANodeAffinity:01 Preferred:true}]
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204941    2011 scope_container.go:77] "TopologyHints" hints=map[cpu:[{NUMANodeAffinity:01 Preferred:true}]] pod="default/testpmd" containerName="dpdk"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204949    2011 config.go:97] "Looking for sources, have seen" sources=[api file] seenSources=map[file:{}]
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204957    2011 scope_container.go:77] "TopologyHints" hints=map[] pod="default/testpmd" containerName="dpdk"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204964    2011 policy.go:78] "Hint Provider has no preference for NUMA affinity with resource" resource="openshift.io/dpdk_nic_1"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204971    2011 policy.go:70] "Hint Provider has no preference for NUMA affinity with any resource"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204979    2011 scope_container.go:85] "ContainerTopologyHint" bestHint={NUMANodeAffinity:01 Preferred:true}
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204987    2011 scope_container.go:54] "Best TopologyHint" bestHint={NUMANodeAffinity:01 Preferred:true} pod="default/testpmd" containerName="dpdk"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.204995    2011 scope_container.go:59] "Topology Affinity" bestHint={NUMANodeAffinity:01 Preferred:true} pod="default/testpmd" containerName="dpdk"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205009    2011 manager.go:914] "Looking for needed resources" needed=1048576000 resourceName="memory"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205016    2011 manager.go:914] "Looking for needed resources" needed=1 resourceName="openshift.io/dpdk_nic_1"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205024    2011 config.go:97] "Looking for sources, have seen" sources=[api file] seenSources=map[file:{}]
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205037    2011 manager.go:694] "Found pre-allocated devices for resource on pod" resourceName="openshift.io/dpdk_nic_1" containerName="dpdk" podUID="c7dcb25a-6b71-4f80-9482-053ea597230c" devices=[0000:00:05.0]
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205044    2011 manager.go:914] "Looking for needed resources" needed=2 resourceName="cpu"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205050    2011 manager.go:914] "Looking for needed resources" needed=1073741824 resourceName="hugepages-1Gi"
Apr 13 09:53:17 TESTNODE hyperkube[2011]: I0413 09:53:17.205060    2011 config.go:97] "Looking for sources, have seen" sources=[api file] seenSources=map[file:{}]
```
When the topology manager requests device allocation, we end up in the devicemanager's [devicesToAllocate](https://github.com/kubernetes/kubernetes/blob/v1.24.0-rc.0/pkg/kubelet/cm/devicemanager/manager.go#L686) function. This function checks if there already a pre-allocated device (and there is, thanks to the recovered data), and if so exits early, skipping the health check which would have failed, because the device plugin has not re-registered itself yet.

The end result is that containers can be admitted and start running without the requested device actually available, causing runtime errors in the workload.

### What did you expect to happen?

The kubelet should reject the admission of the pod until the device plugins re-registered themselves; then the pods requiring devices should be processed again. This seems the only meaningful way to recover pods while taking in consideration the actualy health of the devices.

### How can we reproduce it (as minimally and precisely as possible)?

The behaviour here depends on the order on which pods are recovered, so it's essentially a race.

1. run a test pod requiring devices. Make sure this test pod clearly fails if the requested devices are not available
2. restart the node (can take few attempts)
3. check the health of the test pod

### Kubernetes version

<details>

```console
kubelet 1.23.3
likely happening on current master
```
</details>

### Cloud provider

<details>
bare metal
</details>

### OS version

<details>

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by swatisehgal** [comment]
Changing the priority important-longterm as we should consider the kubelet restart best practices here. We need to look into whether it is reasonable to restart kubelet without draining/cordoning the node.

/priority important-longterm

**Comment by prakharporwal** [comment]
Hi Is this Issue open to work on. I am new to k8s would like to work on this .

**Comment by fromanirh** [comment]
> Hi Is this Issue open to work on. I am new to k8s would like to work on this .

Well this issue require very specific kubelet knowledge so I don't recommend it as first issue

**Comment by prakharporwal** [comment]
@fromanirh I think that's how I will learn . Is this issue time bound ? Can you help me with areas I will have to learn about to work on this.

**Comment by fromanirh** [comment]
> @fromanirh I think that's how I will learn . Is this issue time bound ? Can you help me with areas I will have to learn about to work on this. Else do you have something else I can start with . All the good-first-issue are assigned to someone.

Sure, maybe just have a look to issue tagged "help wanted" and/or "good first issue" to see if there is something friendlier available; if not, and/or if you have interest in this specific area, we can work this out.

**Comment by fromanirh** [comment]
A *possible* workaround, to be evaluated, _could be_ to remove the device manager state file at node boot time, and only at node boot time. We need to keep the state file across kubelet (but not node) reboots.

**Comment by swatisehgal** [comment]
> ### What happened?
> The following behaviour was reproduced on openshift 4.10 ("Kubelet version" kubeletVersion="v1.23.3+54654d2") but I believe the root cause is on kubernetes.

I managed to reproduce this issue on a kind based Kubernetes cluster (v1.24.7) so the root cause is certainly Kubernetes. I Will provide additional updates as I have 'em!

**Comment by swatisehgal** [comment]
I have been spending some time to get a better understanding of this issue and want to capture some of my learning here:
On node restart, here is the flow of events on kubelet:

1. Device Manager [starts](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L246).
2. Assuming we haven't removed the checkpoint file which [readCheckpoint()](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L601) is called using which the following information is populated:

- podDevices (which contains pod to allocated device mapping)
- m.allocatedDevices which stores the list of allocated devices. 
- registered device plugin resources

3. For the registered devices plugins resources obtained in the step above, healthyDevices/unhealthyDevices [are set to the zero value](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L634-L636): (sets.NewString()). 
4. As part of the Node sync loop,  [GetCapacity](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L530) is called where we inspect both healthy (and unhealthy devices). Because of the point above, capacity of of the device plugin resource is update to the length of healthy (and unhealthy) devices present on the system which is 0 in this case as all the devices have been reset. This is done to allow the device plugin to re-register itself so that capacity can be updated appropriately.
5. Assuming a scenario where we have an application pod requesting a device provided by the device plugin pod coming before device plugin pod has had a chance to start or re-register itself, we reach the allocation phase for the application pod: [Allocate() ](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L374) -> [allocateContainerResources](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L902) -> [deviceToAllocate](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L686).
6. In deviceToAllocate, we obtain podDevices which returns the devices that were previously allocated to the application pod  means that the same number of devices are being requested  as were previously allocated  making needed=0 [returning](https://github.com/kubernetes/kubernetes/blob/release-1.24/pkg/kubelet/cm/devicemanager/manager.go#L704) without any obvious error.

I tried https://github.com/kubernetes/kubernetes/pull/110729 but it didn't solve the issue. This is because healthyDevices are reset to empty set of strings keyed by resource name  because of which in the expression `_, ok := m.healthyDevices[resource]`, `ok` returns true as the device plugin had registered itself prior to the node reboot). In order to fix this, we need to  check that we have healthy resources to fulfill the pod request. If we don't have healthy devices, the pod admission should fail. I am working on it and will be testing the fix in the coming days.

**Comment by fromanirh** [comment]
xref: https://github.com/kubernetes/kubernetes/pull/110537
_possibly_ related

**Comment by swatisehgal** [comment]
> xref: #110537 _possibly_ related

Thanks for pointing to this issue, it is definitely related.

**Comment by liggitt** [comment]
reopening since #114640 is reverted in https://github.com/kubernetes/kubernetes/pull/116341

**Comment by smarterclayton** [comment]
This sounds like this adds additional requirements to kubelet admission - the ability to defer for retry admission for some time, rather than rejecting right away?

When you say "reject" are you returning a failure in kubelet admission such that the pod is marked failed?  Or do you mean reject in a different way?

**Comment by ffromani** [comment]
> This sounds like this adds additional requirements to kubelet admission - the ability to defer for retry admission for some time, rather than rejecting right away?

IMO that would be very nice and handy. If kubelet knows it can admit a pod, but not just yet, I *think* that would be a nice addition. Cases I can think of:
1. this very case: pods requesting devices, it the device plugin pod is not recovered yet the application pod should either be postponed or fail admission _and retry later_
2. operation in progress, kubelet knows resources are about to be freed /available soon enough, not yet though

> When you say "reject" are you returning a failure in kubelet admission such that the pod is marked failed? Or do you mean reject in a different way?

I think the best we can have now is a hard reject: pod fails admission and is marked failed. This is what I suggested as desired behavior and what the current fix implements. In general though this can be improved, for example retrying the admission later.

**Comment by smarterclayton** [comment]
Ok, that's useful.  I'll add that to the tracking list for improving admission - the problem of "i can't admit this now but probably can later" comes in with other places (apparmor, waiting for pods to be gracefully deleted).  Thanks

**Comment by swatisehgal** [comment]
> This sounds like this adds additional requirements to kubelet admission - the ability to defer for retry admission for some time, rather than rejecting right away?

I agree that this kind of behavior would be useful but it is worth mentioning that currently we rely on the fact we don't use naked pods and typically a pod is part of a deployment. New instances would be spinned up and at some stage _a_ pod would succeed at admission and that's when the deployment would succeed.

**Comment by ffromani** [comment]
> A gentle ping here. With the fix #116376, if kubelet is restarted on a node, then all the existing and running workloads that use devices are terminated:
> 
> ```
>  Warning  UnexpectedAdmissionError  45s   kubelet            Allocate failed due to no healthy devices present; cannot allocate unhealthy devices devices.kubevirt.io/kvm, which is unexpected
>   Normal   Killing                   42s   kubelet            Stopping container compute
> ```
> 
> This also affects 1.25.x and 1.26.x branches, since the fix has been cherry-picked there.
> 
> I guess this happens because of the race between device plugin registration and the code in `devicesToAllocate`. Anyway, this does not seem right to me. A potential restart of kubelet should not interrupt the running workloads, right? Any thought on that?

Yes, something is not fine here and this deserve discussion, please file a separate issue, backlinking this one, for visibility. Thanks!
